### DISCUSSION TITLE
Position object relative to another object using reference points or reference edges

### DISCUSSION THREAD

**Discussion by WayneSherman** [gumyr/build123d#741] [Q&A] [answered]
In 2D CAD a common workflow to move an object in relation to another object is:

1. Select Object1
2. Select MOVE operation
3. Select source point for the move
4. Select destination point on Object2
5. Object1 gets moved using the chosen reference points.

I would like to position objects in relation to one another with Build123d using reference edges or reference points.  For example, if I have two separate 2D (XY plane) face objects:

![image](https://github.com/user-attachments/assets/565db6aa-ce0b-421a-b68a-7f26039614be)

How can I select the reference vertex points in order to position face1 in relation to face2?

![image](https://github.com/user-attachments/assets/fdb39e52-7904-4f50-9e52-439da81496e1)

or using reference edges, how can one do this?:
1) Move face1 vertically so that face1's inner rectangle top edge is aligned with face2's inner rectangle top edge
and
2) Move face1 horizontally so that face1's inner rectangle left edge is aligned with face2's inner rectangle left edge

**Comment by gumyr** [comment, marked as answer]
Here is an example of how to do this:
```py
from build123d import *
from ocp_vscode import show_all
from random import randrange

# Create two frame objects, one at a random location
frame1 = Rectangle(100, 70) - Rectangle(90, 60)
frame2 = (Rectangle(70, 50) - Rectangle(60, 40)).locate(
    Pos(randrange(0, 100), randrange(0, 100))
)
# Find two alignment points
pt1 = Vector(
    frame1.face().inner_wires()[0].vertices().group_by(Axis.Y)[-1].sort_by(Axis.X)[0]
)
pt2 = Vector(
    frame2.face().inner_wires()[0].vertices().group_by(Axis.Y)[-1].sort_by(Axis.X)[0]
)

# Move frame2 to align with frame1
frame2.position += pt1 - pt2

show_all()
```
Before move:
![image](https://github.com/user-attachments/assets/bbe7c452-4e05-44f8-aac8-55f77a990871)
After move:
![image](https://github.com/user-attachments/assets/29490461-b855-41c6-af6d-f24daa81a6de)

In Builder mode one could with `with Locations(difference_in_position):` to reposition the second object relative to the first.

Hopefully this answers your question.

**Reply by WayneSherman** [reply]
> Here is an example of how to do this:
>  ...
> pt1 = Vector(
>     frame1.face().inner_wires()[0].vertices().group_by(Axis.Y)[-1].sort_by(Axis.X)[0]
> )
> pt2 = Vector(
>     frame2.face().inner_wires()[0].vertices().group_by(Axis.Y)[-1].sort_by(Axis.X)[0]
> )
> 
> #Move frame2 to align with frame1
> frame2.position += pt1 - pt2
>
> In Builder mode one could with `with Locations(difference_in_position):` to reposition the second object relative to the first.
> Hopefully this answers your question.

Thank you for your help, that works for me.  It took me a bit to wrap my head around how group_by and sort_by work, but I think I understand now:

**vertices().group_by(Axis.Y)[-1]**
Group and sort the 4 "inner_wire" vertices into (two) lists with each list containing vertices with the same Y values.
Return the list which has the largest Y values (the last element), which contains the top two vertices.

**.sort_by(Axis.X)[0]**
Sort the two vertices in this list by X values and return the vertex with the smallest X value (the first element), which is the left vertex.

**Reply by gumyr** [reply]
You've got it.  The selectors create Python lists so all of the standard selection and slicing rules apply.
